### ISSUE TITLE
[naming-conventions] React functional component has to be camelcase

### ISSUE THREAD

**Issue by Thaoden** [typescript-eslint/typescript-eslint#2607]
This is similar to issue #2246.

React's naming conventions are to use PascalCase when a function returns a component:

```TypeScript
//Function name `RoleManager` must match one of the following formats: camelCase eslint(@typescript-eslint/naming-convention)
export function RoleManager(): JSX.Element { /* yadda yadda yadda */}
```

The proposed fix in #2246 was to use a rule allowing both `PascalCase` and `camelCase` for all types returning functions; as this here is a function declaration, I've modified the rule to apply to all functions:

```JSON
"@typescript-eslint/naming-convention": [
  "warn",
    {
      "selector": "function",
      "format": ["PascalCase", "camelCase"]
       }
  ]
```

This however allows _all_ functions to be in PascalCase, even though they don't return a React component and thus should effectively be camel cased, eg

```TypeScript
// no warning
const OnClick = () => {/* do nothing */};
```

Is there any guidance as to how to enforce PascalCase for React components without having to be lenient about the casing of all other functions?

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.2.0` |
| `@typescript-eslint/parser`        | `4.2.0` |
| `TypeScript`                       | `4.0.3` |
| `ESLint`                           | `7.10.0` |
| `node`                             | `13.12.0` |

**Comment by bradzacher** [comment]
There's nothing special about a react functional component. It's just a function (that's the beauty of it).

So there's nothing unique to detect and distinguish them from non-react functions.

Which means there's nothing we can do.
If you use a certain naming convention for react files, you could then use eslint overrides to limit the scope of PascalCase, but other than that, sorry.

**Comment by Thaoden** [comment]
That makes sense...

To be honest, I don't know what capabilities the linters have. Would it be possible to check if the returned entity contains a `<>` or a `<div>` or anything like that?

**Comment by bradzacher** [comment]
That could be done. But that's not a true indication of what is a functional component and what is not.

For example I've seen patterns like this a lot:

```ts
function getIcon(val: MyEnum) {
  switch (val) {
    case MyEnum.One:
      return <img src="url" alt="one" />;

      //.......... 
  }
} 

function Component() {
  return (
    <div>
      {getIcon(MyEnum.One)}
      {getIcon(MyEnum.Two)}
      {/*.........*/}
    </div>
  );
}
```

Could `getIcon` be upgraded to a real component? Sure. But it's not defined as such, or used as a component.

You really need:
- returns JSX or null
- accepts 1 argument that is an object, or
- accepts 2 arguments if it's a forwarded ref, or
- accepts 0 arguments

But these are not even correct. Eg you could have a helper function to render jsx (which is impossible to detect without type info). You can track variable references but that's really hard to do well. The arguments are hard to handle because there are different cases.

This is why it's such a tricky thing. A functional component is a component because the engineer decides its a component. 

`eslint-plugin-react` has a whole bunch of code that they use to try and classify functions as components. https://github.com/yannickcr/eslint-plugin-react/blob/master/lib/util/Components.js

**Comment by Thaoden** [comment]
Well, I'll go for both PascalCase and camelcase, then. Thanks for your help!

**Comment by paulmelnikow** [comment]
I'm running into this too. I'd love to be able to exclude function components from this rule.

> There's nothing special about a react functional component. It's just a function (that's the beauty of it).

Would it be feasible to check the return type?

Function components can be defined to return `JSX.Element | JSX.Element[] | null` (or some subset of that).

**Comment by bradzacher** [comment]
> Would it be feasible to check the return type?

There's no guarantee that a function has a return type annotation - which means we would need type information to detect the return type. It's possible to use this (as there is already some type information used by the rule when you use the `types` selector), but it just comes with a cost.

Additionally - there's no guarantee that a function that returns a `JSX.Element` is in fact a react component. Utility functions that return `JSX.Element`, but aren't react components are a relatively common practice, which would cause false-positives.
